components/salsa-2022/src/function/maybe_changed_after.rs

```diff
@@ -170,12 +170,18 @@ where
                 // during this revision or is otherwise stale.
                 return false;
             }
-            QueryOrigin::BaseInput | QueryOrigin::Field => {
-                // BaseInput: This value was `set` by the mutator thread -- ie, it's a base input and it cannot be out of date.
-                // Field: This value is the value of a field of some tracked struct S. It is always updated whenever S is created.
-                // So if a query has access to S, then they will have an up-to-date value.
+            QueryOrigin::BaseInput => {
+                // This value was `set` by the mutator thread -- ie, it's a base input and it cannot be out of date.
                 return true;
             }
+            QueryOrigin::Field => {
```

Comment:
I was thinking one way to get all the tests pass is that we could store the query that created the field (like `QueryOrigin::Assigned`), when we find that a field may be outdated, we can check if the query stack contains the query that created the field, if yes, the old value is valid, otherwise, panic.  

> This probably means we can remove QueryOrigin::Field altogether.

@nikomatsakis  Also, I think we need to keep the `QueryOrigin::Field`, which is handled differently here than `QueryOrigin::Assigned`

Reply:
Ah, hmm, I see. Yes. But actually it doesn't  *have* to be handled differently here. I think returning false is actually better (panicking in this code makes me a bit nervous, to be honest, although I think it should be fine).

The end-result will be that we try to execute the function associated with the field, which will panic. 

Basically by returning false we're saying: we can't verify that the value in this function is up-to-date.